src/Database.php

```diff
@@ -268,36 +277,63 @@ public function createCollection(string $collectionName, array $options = [])
 
-            $encryptedFields = (array) $encryptedFields;
-            $enxcolOptions = ['clusteredIndex' => ['key' => ['_id' => 1], 'unique' => true]];
-            (new CreateCollection($this->databaseName, $encryptedFields['escCollection'] ?? 'enxcol_.' . $collectionName . '.esc', $enxcolOptions))->execute($server);
-            (new CreateCollection($this->databaseName, $encryptedFields['eccCollection'] ?? 'enxcol_.' . $collectionName . '.ecc', $enxcolOptions))->execute($server);
-            (new CreateCollection($this->databaseName, $encryptedFields['ecocCollection'] ?? 'enxcol_.' . $collectionName . '.ecoc', $enxcolOptions))->execute($server);
-        }
+        $server = select_server($this->manager, $options);
 
-        $operation = new CreateCollection($this->databaseName, $collectionName, $options);
+        return $operation->execute($server);
+    }
 
-        $result = $operation->execute($server);
+    /**
+     * Create a new encrypted collection explicitly.
+     *
+     * The "encryptedFields" option is required.
+     *
+     * This method will automatically create data keys for any encrypted fields
+     * where "keyId" is null. A copy of the modified "encryptedFields" option
+     * will be returned in addition to the result from creating the collection.
+     *
+     * If any error is encountered creating data keys or the collection, a
+     * CreateEncryptedCollectionException will be thrown. The original exception
+     * and modified "encryptedFields" option can be accessed via the
+     * getPrevious() and getEncryptedFields() methods, respectively.
+     *
+     * @see CreateCollection::__construct() for supported options
+     * @return array A tuple containing the command result document from creating the collection and the modified "encryptedFields" option
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws CreateEncryptedCollectionException for any errors creating data keys or creating the collection
+     */
+    public function createEncryptedCollection(string $collectionName, ClientEncryption $clientEncryption, string $kmsProvider, ?array $masterKey, array $options): array
+    {
+        if (! isset($options['typeMap'])) {
+            $options['typeMap'] = $this->typeMap;
+        }
 
-        if ($encryptedFields !== null) {
-            (new CreateIndexes($this->databaseName, $collectionName, [['key' => ['__safeContent__' => 1]]]))->execute($server);
+        if (! isset($options['writeConcern']) && ! is_in_transaction($options)) {
+            $options['writeConcern'] = $this->writeConcern;
         }
 
-        return $result;
+        $operation = new CreateEncryptedCollection($this->databaseName, $collectionName, $options);
+        $server = select_server($this->manager, $options);
+
+        try {
+            $operation->createDataKeys($clientEncryption, $kmsProvider, $masterKey, $encryptedFields);
```

Comment:
I originally implemented _two_ operation classes:

 * CreateCollectionForQueryableEncryption for `createCollection()` with an `encryptedFields` option, which composed several CreateCollection operations and one CreateIndexes operation.
 * CreateEncryptedCollection for this helper, which composed a CreateCollectionForQueryableEncryption.

In that approach, `CreateEncryptedCollection::execute()` created data keys before calling `CreateCollectionForQueryableEncryption::execute()` and contained the logic for CreateEncryptedCollectionException. The downside was updating the options for the inner-most CreateCollection operation when `encryptedFields` changed.

I also wasn't too happy with the naming for both operations. It made sense for CreateEncryptedCollection to match the helper method, but it was only responsible for creating data keys. CreateCollectionForQueryableEncryption was actually creating an encrypted collection.

----

I then combined both operations into CreateEncryptedCollection and moved the key generation logic to a separate `createDataKeys()` method to be called prior to `execute()`. This made it easier to reconstruct the inner CreateCollection operation.

Using a pass-by-reference parameter to expose modifications to `encryptedFields` allowed me to avoid dealing with CreateEncryptedCollectionException in the operation class entirely. I briefly explored returning `encryptedFields` _and_ throwing the exception when needed, and it made both `createDataKeys()` and the helper's `try/catch` more complicated.